Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -1,217 +1,269 @@
+  which could be viewed as a superset of `x86`.
+  + The logical grouping of `ARM64` and `ARM32` instruction sets is different.  It is controlled by
+  different sets of `System Registers`.
 
-+ `System.Runtime.Intrinsics.Arm.ARMv7` is used for `ARMv7` added functionality not included in `Base`
+For the convenience of the end user, it may be useful to add convenience API's which expose functionality
+which is common across platforms and sets of platforms.  These could be implemented in terms of the
+platform specific functionality.  These API's are currently out of scope of this initial design document.
 
-If an extension is only supported on `Arm64`, there is no reason to add a sub namespace.  Hypothetically:
+### Logical Set Class Names
 
-+ `System.Runtime.Intrinsics.Arm.ARMv10` is used for `ARMv10` added functionality (Hypothetically `Arm64` only)
+Within the `System.Runtime.Intrinsics.Arm.Arm64` namespace there will be a separate `static class` for each
+logical set of instructions
 
-Namespace is chosen based on what could be implemented, not on what is or will be implemented.
+The sets will be chosen to match the granularity if the `ARM64` `ID_*` register fields.
 
-#### Names
+#### Specific Class Names
 
-Intrinsics will be named to describe functionality.  Names will not correspond to specific named assembly instructions.
+TBD Add table to document each known extension and recommended `static class` name.
 
-`IsSupported` will be used in any namespace which may not be implemented on all CoreCLR platforms
+### Intrinsic Method Names
 
-### Intrinsic Interface Documentation
+Intrinsics will be named to describe functionality.  Names will not correspond to specific named
+assembly instructions.
 
-+ Documentation will be minimal preferring the underlying ARM documentation
-+ Namespaces will briefly document corresponding specification
-+ Intrinsic methods will briefly document corresponding assembly instruction(s) for each platform
+Where precedence exists within the `System.Runtime.Intrinsics.X86` namespace, identical method names will be
+chosen: `Add`, `Multiply`, `Load`, `Store` ...
 
-### Test coverage
+### Intinsic Method Argument Types
 
-As intrinsic support is added test coverage must be extended to provide basic testing
+Intrinsic methods will typically use a standard set of argument types:
++ Integer type: `byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`
++ Floating types: `double`, `single`, `System.Half`
++ Vector types: Vector128<T>
++ SVE will add new vector types: TBD
+
+## Intrinsic Interface Documentation
 
-### Phased Implementation
++ Namespace
++ Each `static class` will
+  + Briefly document corresponding `System Register Field and Value` from ARM specification.
+  + Document use of IsSupported property
+  + Optionally summarize set of methods enabled by the extension
++ Each intrinsic method will
+  + Document underlying `ARM64` assembly instruction
```

Comment:
It may be worth also documenting the corresponding compiler intrinsic, where applicable.

The MSVC compiler keeps a documented list of those names here: https://msdn.microsoft.com/en-us/library/hh875058.aspx.

I do not know if GCC/Clang (or other compilers) follow a similar naming convention for their ARM intrinsics.

Reply:
I'll add a note, but `ARM64` intrinsics are almost non-existant.